pulsar-broker/src/main/java/org/apache/pulsar/broker/rest/TopicsBase.java

```diff
@@ -0,0 +1,733 @@
+                messageMetadata.addAllReplicateTos(message.getReplicationClusters());
+            }
+
+            if (null != message.getProperties()) {
+                messageMetadata.addAllProperties(message.getProperties().entrySet().stream().map(entry -> {
+                    org.apache.pulsar.common.api.proto.KeyValue keyValue =
+                            new org.apache.pulsar.common.api.proto.KeyValue();
+                    keyValue.setKey(entry.getKey());
+                    keyValue.setValue(entry.getValue());
+                    return keyValue;
+                }).collect(Collectors.toList()));
+            }
+            if (null != message.getKey()) {
+                // If has key schema, encode partition key, else use plain text.
+                if (schema.getSchemaInfo().getType() == SchemaType.KEY_VALUE) {
+                    KeyValueSchemaImpl kvSchema = (KeyValueSchemaImpl) schema;
+                    messageMetadata.setPartitionKey(
+                            Base64.getEncoder().encodeToString(encodeWithSchema(message.getKey(),
+                                    kvSchema.getKeySchema())));
+                    messageMetadata.setPartitionKeyB64Encoded(true);
+                } else {
+                    messageMetadata.setPartitionKey(message.getKey());
+                    messageMetadata.setPartitionKeyB64Encoded(false);
+                }
+            }
+            if (null != message.getEventTime() && !message.getEventTime().isEmpty()) {
+                messageMetadata.setEventTime(Long.valueOf(message.getEventTime()));
+            }
+            if (message.isDisableReplication()) {
+                messageMetadata.clearReplicateTo();
+                messageMetadata.addReplicateTo("__local__");
+            }
+            if (message.getDeliverAt() != 0 && messageMetadata.hasEventTime()) {
+                messageMetadata.setDeliverAtTime(message.getDeliverAt());
+            } else if (message.getDeliverAfterMs() != 0) {
+                messageMetadata.setDeliverAtTime(messageMetadata.getEventTime() + message.getDeliverAfterMs());
+            }
+            if (schema.getSchemaInfo().getType() == SchemaType.KEY_VALUE) {
+                KeyValueSchemaImpl kvSchema = (KeyValueSchemaImpl) schema;
+                pulsarMessages.add(MessageImpl.create(messageMetadata,
+                        ByteBuffer.wrap(encodeWithSchema(message.getPayload(), kvSchema.getValueSchema())),
+                        schema, topicName.toString()));
+            } else {
+                pulsarMessages.add(MessageImpl.create(messageMetadata,
+                        ByteBuffer.wrap(encodeWithSchema(message.getPayload(), schema)), schema,
+                        topicName.toString()));
+            }
+        }
+
+        return pulsarMessages;
+    }
+
+    // Encode message with corresponding schema, do necessary conversion before encoding
+    private byte[] encodeWithSchema(String input, Schema schema) {
```

Comment:
Should it be moved to SchemaUtils?

Reply:
I took a look, and I think this method is very specific for REST API, mostly encode a message in String with given schema and contains some custom logic when handling Json/Avro schema, might not be useful for other use case, so I think we can keep it here for now, and if we find it can we reuse somewhere else then we can move it to SchemaUtils later.